File: sdk/examples/nodejs/quickstart.mjs
/**
 * Quickstart for the published `rfb-sdk` package (npm).
 *
 * Prerequisites:
 *   * a running forkd controller (default http://127.0.0.1:8889, or set
 *     FORKD_URL / FORKD_TOKEN);
 *   * a ready + bootable snapshot, e.g. created with
 *     `rfb-cli forkd snapshot-create --tag rfb --tap forkd-tap0`.
 *
 * Install: `npm install rfb-sdk`
 * Run:     `node quickstart.mjs rfb`
 */
import { RfbClient, RfbError } from 'rfb-sdk';

const tag = process.argv[2] ?? 'rfb';

try {
  // FORKD_URL / FORKD_TOKEN / 10s timeout are the defaults.
  const client = new RfbClient();

  // Block until the snapshot reports status=ready and bootable=true.
  const snapshot = await client.waitSnapshot(tag);
  console.log(`snapshot ${snapshot.tag} is ready`);

  const [sandbox] = await client.createSandbox(tag);
  console.log(`sandbox ${sandbox.id} created`);

  console.log('ping:', await sandbox.ping());

  const result = await sandbox.exec(['echo', 'hello'], { cwd: '/workspace' });
  console.log(`exec: exit=${result.exitCode} stdout=${result.stdoutText.trim()}`);

  const written = await sandbox.write('notes.txt', 'hello from rfb-sdk');
  console.log('written:', written, 'bytes');

  const file = await sandbox.read('notes.txt');
  console.log('read back', file.data.length, 'bytes');

  const entries = await sandbox.ls('/workspace');
  console.log('ls:', entries.map((entry) => entry.name));

  await sandbox.delete();
  console.log('sandbox deleted');
} catch (error) {
  if (error instanceof RfbError) {
    console.error(`rfb error: ${error.message}`);
    process.exit(1);
  }
  throw error;
}

File: sdk/examples/python/quickstart.py
"""Quickstart for the published `rfb-sdk` package (PyPI).

Prerequisites:
  * a running forkd controller (default http://127.0.0.1:8889, or set
    FORKD_URL / FORKD_TOKEN);
  * a ready + bootable snapshot, e.g. created with
    `rfb-cli forkd snapshot-create --tag rfb --tap forkd-tap0`.

Install: `pip install rfb-sdk`
Run:     `python quickstart.py rfb`
"""

import sys

from rfb_sdk import RfbClient, RfbError


def main() -> int:
    tag = sys.argv[1] if len(sys.argv) > 1 else "rfb"

    # FORKD_URL / FORKD_TOKEN / 10s timeout are the defaults.
    client = RfbClient()

    # Block until the snapshot reports status=ready and bootable=true.
    snapshot = client.wait_snapshot(tag)
    print(f"snapshot {snapshot.tag} is ready")

    sandbox = client.create_sandbox(tag)[0]
    print(f"sandbox {sandbox.id} created")

    print("ping:", sandbox.ping())

    result = sandbox.exec(["echo", "hello"], cwd="/workspace")
    print("exec: exit=%d stdout=%s" % (result.exit_code, result.stdout_text.strip()))

    written = sandbox.write("notes.txt", b"hello from rfb-sdk")
    print("written:", written, "bytes")

    file = sandbox.read("notes.txt")
    print("read back", len(file.data), "bytes")

    print("ls:", [entry.name for entry in sandbox.ls("/workspace")])

    sandbox.delete()
    print("sandbox deleted")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except RfbError as error:
        print(f"rfb error: {error}", file=sys.stderr)
        raise SystemExit(1)
